### ISSUE TITLE
ttnn.selu: the registered golden ignores the op's own scale= and alpha= kwargs, returning standard SELU — up to 3416% off while PCC stays above 0.99

### ISSUE THREAD

**Issue by p32929** [tenstorrent/tt-metal#57116]
### Component / Area

`ttnn` eltwise unary goldens — `_golden_function_selu` in `ttnn/ttnn/operations/unary.py`, registered onto `ttnn.selu`.

### Issue Type (optional)

Bad Outputs

### Observed

`ttnn.selu` takes `scale` and `alpha` as real keyword arguments. Its registered golden discards both:

```python
# ttnn/ttnn/operations/unary.py @ a4e9bec
def _golden_function_selu(input_tensor_a, *args, **kwargs):
    import torch

    return torch.nn.functional.selu(input_tensor_a)


ttnn.attach_golden_function(ttnn.selu, golden_function=_golden_function_selu)
```

`torch.nn.functional.selu` has no `scale`/`alpha` parameters at all — it hardcodes the canonical constants — so whatever the caller passed lands in `**kwargs` and is thrown away. The golden returns the identical tensor for `scale=3.0, alpha=2.0` and for `scale=0.5, alpha=0.1`.

Both parameters are genuinely user-settable and are threaded all the way into the kernel:

```cpp
// ttnn/cpp/ttnn/operations/eltwise/unary/unary_nanobind.cpp:1999-2000
nb::arg("scale") = 1.0507f,
nb::arg("alpha") = 1.67326f,

// ttnn/cpp/ttnn/operations/eltwise/unary/unary.cpp:267
DEFINE_UNARY_OP_WITH_TWO_FLOAT_PARAMS(selu, SELU)
//   -> UnaryWithParam{UnaryOpType::SELU, {parameter_a, parameter_b}}

// tt_metal/hw/ckernels/wormhole_b0/metal/llk_api/llk_sfpu/ckernel_sfpu_selu.h
// selu(x) = scale * x for x>=0, scale * alpha * (exp(x)-1) for x<0
const sfpi::vFloat scale_alpha = Converter::as_float(scale) * Converter::as_float(alpha);
sfpi::vFloat result = scale_alpha * expm1_cw_clamped(x);
v_if(x >= 0.0f) { result = scale_val * x; }
```

`ttnn/ttnn/decorators.py:1763` (and `:1375`/`:1399`) forward the caller's `**kwargs` to the golden verbatim, so `scale=`/`alpha=` do arrive — they are simply dropped.

**`selu` is the only op in its family with this gap.** Every sibling two-parameter unary names its parameters in the golden signature and forwards them:

| op | golden signature |
|---|---|
| `softplus` | `_golden_function_softplus(input_tensor_a, beta, threshold, ...)` |
| `hardtanh` | `_golden_function_hardtanh(input_tensor_a, min_val, max_val, ...)` |
| `threshold` | `_golden_function_threshold(input_tensor, threshold, value, ...)` |
| `logit` | `_golden_function_logit(input_tensor_a, eps, ...)` |
| `clip` / `clamp` | both take `(min, max)` |

The only existing coverage, `tests/ttnn/unit_tests/operations/eltwise/test_eltwise_selu.py`, only ever calls `ttnn.selu(x)` with defaults and asserts at PCC 0.99, so it cannot see this.

**Not the same as #55393.** That issue (and PR #55565) is about the *truncated default constants* — a 43 ULP bias on the default path, fixed in the nanobind and in `selu_bw`. Neither touches `ttnn/ttnn/operations/unary.py`, so after #55565 merges the golden will still ignore user-supplied `scale`/`alpha` entirely. Row 2 of the table below is #55393, shown only to keep the two separated.

### Expected

The golden should evaluate the same two-parameter formula the kernel does, reducing to `torch.nn.functional.selu` at the canonical defaults.

### Magnitude

Pure Python, no hardware, no ttnn build. `(1,1,32,32)` float32, `uniform(-8, 8)`, `torch.manual_seed(0)`:

```
torch 2.8.0   input (1,1,32,32) float32 uniform(-8,8), manual_seed(0)
(507 negative / 517 non-negative elements)

ttnn.selu(x, scale=, alpha=)                    max rel err   mean rel err   caught by PCC >= 0.99?
---------------------------------------------------------------------------------------------------
canonical defaults (as merged by PR #55565)         0.0000%        0.0000%       NO  (PCC 1.000000)
current main's truncated defaults (#55393)          0.0003%        0.0002%       NO  (PCC 1.000000)
scale=1.0, alpha=1.0  (plain ELU)                  75.8099%       40.0946%       NO  (PCC 0.998291)
scale=2.0, alpha=1.67326                           47.4650%       47.4649%       NO  (PCC 1.000000)
scale=1.0507, alpha=0.5                           234.6530%      116.1808%       NO  (PCC 0.994096)
scale=3.0, alpha=2.0                               70.6983%       67.8096%       NO  (PCC 0.999685)
scale=0.5, alpha=0.1                             3416.1988%     1747.0265%       yes (PCC 0.988194)
```

Row 1 is the control: at the canonical constants the device model and the golden agree **bit-for-bit**, which is what makes the rest attributable to the dropped parameters rather than to my transcription of the kernel.

The error is a multiplicative bias of `scale / 1.0507` on the positive branch and `scale * alpha / 1.7581` on the negative one, so it is **unbounded in `scale * alpha`** — not an edge case and not a rounding effect.

The last column is the part worth noting: SELU's two branches are both linear-ish in the same direction, so re-weighting them barely decorrelates the output. **Six of the seven cases — including one with a 1747% mean relative error — still pass a `PCC >= 0.99` gate.** Any PCC-based comparison against this golden reports a pass while the reference is wrong by more than an order of magnitude.

### Reproduction Steps

**1. Steps (exact commands)**

```bash
pip install torch          # only dependency; 2.8.0 used above
python repro_selu_golden_ignores_scale_alpha.py
```

Ends in an `assert`-based self-check that (a) pins the control row bit-exact, so the device model is validated before any divergence is claimed, and (b) asserts `torch.equal` between the golden's output at `scale=3.0, alpha=2.0` and at `scale=0.5, alpha=0.1` — a direct demonstration that the kwargs are ignored.

<details>
<summary><code>repro_selu_golden_ignores_scale_alpha.py</code></summary>

```python
#!/usr/bin/env python3
"""
Repro: the registered golden for `ttnn.selu` ignores the op's own `scale=` and
`alpha=` keyword arguments.

    # ttnn/ttnn/operations/unary.py @ a4e9bec
    def _golden_function_selu(input_tensor_a, *args, **kwargs):
        import torch
        return torch.nn.functional.selu(input_tensor_a)   # no scale, no alpha

`torch.nn.functional.selu` takes no `scale`/`alpha` at all -- it hardcodes the
canonical SELU constants -- so whatever the caller passed is discarded.

No hardware needed. The device model below is transcribed verbatim from the SFPU
kernel:

    // tt_metal/hw/ckernels/wormhole_b0/metal/llk_api/llk_sfpu/ckernel_sfpu_selu.h
    // selu(x) = scale * x for x>=0, scale * alpha * (exp(x)-1) for x<0
    const sfpi::vFloat scale_alpha = Converter::as_float(scale) * Converter::as_float(alpha);
    sfpi::vFloat result = scale_alpha * expm1_cw_clamped(x);
    v_if(x >= 0.0f) { result = scale_val * x; }

and both parameters really are user-settable:

    // ttnn/cpp/ttnn/operations/eltwise/unary/unary_nanobind.cpp:1999-2000
    nb::arg("scale") = 1.0507f,
    nb::arg("alpha") = 1.67326f,
    // unary.cpp:267  DEFINE_UNARY_OP_WITH_TWO_FLOAT_PARAMS(selu, SELU)
    //   -> UnaryWithParam{UnaryOpType::SELU, {parameter_a, parameter_b}}

Run:  python repro_selu_golden_ignores_scale_alpha.py
"""
import torch

torch.manual_seed(0)

CANON_SCALE = 1.0507009873554805   # torch.nn.functional.selu's constants
CANON_ALPHA = 1.6732632423543772


# ---- golden under test, VERBATIM from ttnn/ttnn/operations/unary.py ----------
def _golden_function_selu(input_tensor_a, *args, **kwargs):
    return torch.nn.functional.selu(input_tensor_a)


# ---- device model, transcribed from ckernel_sfpu_selu.h ---------------------
def device_selu(x, scale, alpha):
    scale32 = torch.tensor(scale, dtype=torch.float32)
    alpha32 = torch.tensor(alpha, dtype=torch.float32)
    scale_alpha = scale32 * alpha32
    return torch.where(x >= 0.0, scale32 * x, scale_alpha * torch.expm1(x))


def max_rel_err_pct(dev, gold):
    dev, gold = dev.double(), gold.double()
    return ((dev - gold).abs() / dev.abs().clamp_min(1e-30)).max().item() * 100.0


def mean_rel_err_pct(dev, gold):
    dev, gold = dev.double(), gold.double()
    return ((dev - gold).abs() / dev.abs().clamp_min(1e-30)).mean().item() * 100.0


x = torch.empty(1, 1, 32, 32, dtype=torch.float32).uniform_(-8.0, 8.0)
neg, pos = x[x < 0], x[x >= 0]
print(f"torch {torch.__version__}   input (1,1,32,32) float32 uniform(-8,8), manual_seed(0)")
print(f"({neg.numel()} negative / {pos.numel()} non-negative elements)\n")

CASES = [
    ("canonical defaults (as merged by PR #55565)", CANON_SCALE, CANON_ALPHA),
    ("current main's truncated defaults (#55393)", 1.0507, 1.67326),
    ("scale=1.0, alpha=1.0  (plain ELU)", 1.0, 1.0),
    ("scale=2.0, alpha=1.67326", 2.0, 1.67326),
    ("scale=1.0507, alpha=0.5", 1.0507, 0.5),
    ("scale=3.0, alpha=2.0", 3.0, 2.0),
    ("scale=0.5, alpha=0.1", 0.5, 0.1),
]

hdr = "%-44s %14s %14s %24s" % ("ttnn.selu(x, scale=, alpha=)", "max rel err", "mean rel err", "caught by PCC >= 0.99?")
print(hdr)
print("-" * len(hdr))


def pcc(a, b):
    a, b = a.flatten().double(), b.flatten().double()
    a, b = a - a.mean(), b - b.mean()
    return (a @ b / (a.norm() * b.norm())).item()


rows = []
for label, s, a in CASES:
    dev = device_selu(x, s, a)
    gold = _golden_function_selu(x, scale=s, alpha=a)
    r, m, p = max_rel_err_pct(dev, gold), mean_rel_err_pct(dev, gold), pcc(dev, gold)
    rows.append((label, s, a, r, m, p))
    print("%-44s %13.4f%% %13.4f%% %24s" % (label, r, m, ("NO  (PCC %.6f)" % p) if p >= 0.99 else ("yes (PCC %.6f)" % p)))

print("""
Reading of the table
--------------------
Row 1 is the control: at the canonical constants the device model and the golden
agree bit-for-bit, which proves the device model transcribed above is
correct and that everything below is attributable to the dropped parameters.

Row 2 is the already-filed #55393 (truncated Python defaults, fixed by PR
#55565) -- out of scope here, shown only to separate it from this defect.

Rows 3-7 are this bug: the caller passed `scale=`/`alpha=`, the kernel honoured
them, and the golden silently returned standard SELU. The error is a
multiplicative bias of `scale/1.0507` on the positive branch and
`scale*alpha/1.7581` on the negative one, so it is unbounded in `scale*alpha` --
not an edge case and not a rounding issue.

Note that `selu` is the ONLY op in its family with this gap. Every sibling
two-parameter unary names its parameters in the golden signature and forwards
them:

    softplus  -> _golden_function_softplus(input_tensor_a, beta, threshold, ...)
    hardtanh  -> _golden_function_hardtanh(input_tensor_a, min_val, max_val, ...)
    threshold -> _golden_function_threshold(input_tensor, threshold, value, ...)
    logit     -> _golden_function_logit(input_tensor_a, eps, ...)
    clip / clamp -> both take (min, max)

Suggested fix (mirrors the siblings exactly):

    def _golden_function_selu(input_tensor_a, *args, scale=1.0507009873554805,
                              alpha=1.6732632423543772, **kwargs):
        import torch
        return torch.where(input_tensor_a >= 0,
                           scale * input_tensor_a,
                           scale * alpha * torch.expm1(input_tensor_a))

which reduces to torch.nn.functional.selu at the defaults.
""")

# --------------------------------------------------------------- self-check
ctrl = CASES[0]
assert torch.equal(device_selu(x, ctrl[1], ctrl[2]), _golden_function_selu(x)), \
    "control must be bit-exact -- otherwise the device model is wrong"
for label, s, a, r, m, p in rows[2:]:
    assert r > 10.0, f"{label}: expected a large divergence, got {r}%"
assert max_rel_err_pct(device_selu(x, 3.0, 2.0), _golden_function_selu(x, scale=3.0, alpha=2.0)) > 60.0, \
    "scale=3.0/alpha=2.0 must be >60% off"
# the golden is provably insensitive to its own kwargs
a1 = _golden_function_selu(x, scale=3.0, alpha=2.0)
a2 = _golden_function_selu(x, scale=0.5, alpha=0.1)
assert torch.equal(a1, a2), "golden output must be identical for different scale/alpha -- the kwargs are ignored"
print("self-check: OK  (golden returns the identical tensor for scale=3.0,alpha=2.0 and scale=0.5,alpha=0.1)")

```

</details>

**2. Input data / link or description**

Generated in-script: `torch.empty(1, 1, 32, 32, dtype=torch.float32).uniform_(-8.0, 8.0)` under `torch.manual_seed(0)` — 507 negative and 517 non-negative elements, so both kernel branches are exercised.

**3. Frequency**

100% — deterministic, every call that passes a non-default `scale` or `alpha`.

### Environment

**1. Software Versions**

`main` @ `a4e9bec`. torch 2.8.0, Python 3.9.6, macOS arm64. No device needed — the affected code is host-side Python.

**2. Hardware Details**

N/A. The defect is in the host-side golden reference, not a kernel.

### Is this a regression?

No — present since `_golden_function_selu` was registered, and unaffected by the in-flight #55393 / PR #55565 work.

### Impact

Comparison mode, `ttnn.get_golden_function(ttnn.selu)`, and any sweep or fallback that leans on the registered golden all get a wrong reference for any parameterised `selu` call. Because SELU's branches stay highly correlated under re-weighting, PCC gates do not catch it, so a genuine kernel regression in the parameterised path would be masked rather than surfaced. #57086 (Category 6 — exhaustive BF16 coverage for ops with multiple/special parameters) is exactly the kind of suite that would compare parameterised `selu` against this golden.

### Suggested fix

Three lines, mirroring what the sibling ops already do:

```python
def _golden_function_selu(
    input_tensor_a, *args, scale=1.0507009873554805, alpha=1.6732632423543772, **kwargs
):
    import torch

    return torch.where(
        input_tensor_a >= 0,
        scale * input_tensor_a,
        scale * alpha * torch.expm1(input_tensor_a),
    )
```

This is bit-exact with `torch.nn.functional.selu` at the defaults (row 1 above), so it cannot regress the existing default-path test.

Happy to open the PR, together with a unit test that pins a couple of non-default `(scale, alpha)` pairs against the registered golden rather than against a hand-rolled expression — if that is useful.

**Comment by p32929** [comment]
PR up: #57143.

Names `scale`/`alpha` on `_golden_function_selu` and applies them, mirroring `_golden_function_softplus` in the same file. `+9 / -2`, one file.

The default path is short-circuited back to `torch.nn.functional.selu`, so it stays **bit-identical** — `scale * elu(x, alpha)` with the full-precision constants differs from fused `F.selu` by up to 1 fp32 ulp on ~14.5% of elements, and I did not want a no-op call to drift. Only an explicitly parameterised call takes the new path.

Deliberately does not touch the truncated `1.0507f`/`1.67326f` binding defaults — that is #55393 / #55565's scope.

Verified with a device-free torch script: default `torch.equal` to `F.selu`; parameterised `torch.equal` to `scale * F.elu(x, alpha=alpha)` for (1.0507, 1.67326), (3.0, 2.0), (0.5, 0.1), (2.0, 1.0); and `golden(x, scale=3.0, alpha=2.0) != golden(x, scale=0.5, alpha=0.1)` now holds. Say the word and I will add a parametrised unit test in the same PR.
